File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_05/SwitchMapTest.java
package com.fResult.reactor.ch5_05;

import java.time.Duration;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

class SwitchMapTest {
  @Test
  void switchMapWithLookahead() {
    val source =
        Flux.just("re", "rea", "reac", "react", "reacti", "reactiv", "reactive")
            .delayElements(Duration.ofMillis(100))
            .switchMap(this::lookup);

    StepVerifier.create(source)
        .expectNext("reactive -> reactive")
        .expectComplete()
        .verify(Duration.ofSeconds(5));
  }

  private Flux<String> lookup(String word) {
    return Flux.just(word + " -> reactive").delayElements(Duration.ofMillis(500));
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_05/SwitchMapTest.kt
package com.fResult.reactor.ch5_05

import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import kotlin.test.Test
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.toJavaDuration

class SwitchMapTest {
  @Test
  fun switchMapWithLookaheads() {
    val source = Flux.just("re", "rea", "react", "reactive")
      .delayElements(100L.milliseconds.toJavaDuration())
      .switchMap(::lookup)

    StepVerifier.create(source)
      .expectNext("reactive -> reactive")
      .verifyComplete()
  }

  private fun lookup(word: String): Flux<String> {
    return Flux.just("$word -> reactive").delayElements(500L.milliseconds.toJavaDuration())
  }
}
